source/endpoints_ruleset.c

```diff
@@ -0,0 +1,939 @@
+        return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+    }
+
+    size_t num_rules = aws_json_get_array_size(rules_node);
+    aws_array_list_init_dynamic(&rule_data->rules, allocator, num_rules, sizeof(struct aws_endpoints_rule));
+    if (s_init_array_from_json(allocator, rules_node, &rule_data->rules, s_on_rule_element)) {
+        aws_endpoints_rule_data_tree_clean_up(rule_data);
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to parse rules.");
+        return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_on_rule_element(
+    size_t idx,
+    const struct aws_json_value *value,
+    bool *out_should_continue,
+    void *user_data) {
+    (void)idx;
+    (void)out_should_continue;
+    struct array_parser_wrapper *wrapper = user_data;
+
+    /* Required fields */
+    struct aws_byte_cursor type_cur;
+    struct aws_json_value *type_node = aws_json_value_get_from_object(value, aws_byte_cursor_from_c_str("type"));
+    if (type_node == NULL || aws_json_value_get_string(type_node, &type_cur)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract rule type.");
+        return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+    }
+
+    enum aws_endpoints_rule_type type;
+    if (aws_byte_cursor_eq_ignore_case(&type_cur, &s_endpoint_type_cur)) {
+        type = AWS_ENDPOINTS_RULE_ENDPOINT;
+    } else if (aws_byte_cursor_eq_ignore_case(&type_cur, &s_error_type_cur)) {
+        type = AWS_ENDPOINTS_RULE_ERROR;
+    } else if (aws_byte_cursor_eq_ignore_case(&type_cur, &s_tree_type_cur)) {
+        type = AWS_ENDPOINTS_RULE_TREE;
+    } else {
+        AWS_FATAL_ASSERT(false);
+    }
+
+    struct aws_endpoints_rule rule;
+    AWS_ZERO_STRUCT(rule);
+    rule.type = type;
+
+    struct aws_json_value *conditions_node =
+        aws_json_value_get_from_object(value, aws_byte_cursor_from_c_str("conditions"));
+    if (conditions_node == NULL || !aws_json_value_is_array(conditions_node)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Conditions node missing.");
+        goto error_clean_up;
+    }
+
+    size_t num_conditions = aws_json_get_array_size(conditions_node);
+    aws_array_list_init_dynamic(
+        &rule.conditions, wrapper->allocator, num_conditions, sizeof(struct aws_endpoints_condition));
+
+    if (s_init_array_from_json(wrapper->allocator, conditions_node, &rule.conditions, s_on_condition_element)) {
+        AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract conditions.");
+        goto error_clean_up;
+    }
+
+    switch (type) {
+        case AWS_ENDPOINTS_RULE_ENDPOINT: {
+            struct aws_json_value *endpoint_node =
+                aws_json_value_get_from_object(value, aws_byte_cursor_from_c_str("endpoint"));
+            if (s_parse_endpoints_rule_data_endpoint(wrapper->allocator, endpoint_node, &rule.rule_data.endpoint)) {
```

Comment:
we need to properly null-check `endpoint_node` here, or in `s_parse_endpoints_rule_data_endpoint()`